src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/DiagnosticSourceManager.cs

```diff
@@ -0,0 +1,133 @@
+{
+    private readonly ImmutableDictionary<string, IDiagnosticSourceProvider> _documentProviders;
+    private readonly ImmutableDictionary<string, IDiagnosticSourceProvider> _workspaceProviders;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public DiagnosticSourceManager([ImportMany] IEnumerable<IDiagnosticSourceProvider> sourceProviders)
+    {
+        _documentProviders = sourceProviders
+                .Where(p => p.IsDocument)
+                .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+
+        _workspaceProviders = sourceProviders
+                .Where(p => !p.IsDocument)
+                .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<string> GetSourceNames(bool isDocument)
+        => (isDocument ? _documentProviders : _workspaceProviders).Keys;
+
+    /// <inheritdoc />
+    public async ValueTask<ImmutableArray<IDiagnosticSource>> CreateDiagnosticSourcesAsync(RequestContext context, string? sourceName, bool isDocument, CancellationToken cancellationToken)
+    {
+        var providersDictionary = isDocument ? _documentProviders : _workspaceProviders;
+        if (sourceName != null)
+        {
+            // VS does not distinguish between document and workspace sources. Thus it can request
+            // document diagnostics with workspace source name. We need to handle this case.
+            if (providersDictionary.TryGetValue(sourceName, out var provider))
+                return await provider.CreateDiagnosticSourcesAsync(context, cancellationToken).ConfigureAwait(false);
+
+            context.TraceInformation($"No `{sourceName}` diagnostics for {isDocument}");
+            return [];
+        }
+        else
+        {
+            // VS Code (and legacy VS ?) pass null sourceName when requesting all sources.
+            using var _ = ArrayBuilder<IDiagnosticSource>.GetInstance(out var sourcesBuilder);
+            foreach (var kvp in providersDictionary)
+            {
+                // Exclude Task diagnostics from the aggregated sources.
+                if (kvp.Key != PullDiagnosticCategories.Task)
+                {
+                    var namedSources = await kvp.Value.CreateDiagnosticSourcesAsync(context, cancellationToken).ConfigureAwait(false);
+                    sourcesBuilder.AddRange(namedSources);
+                }
+            }
+
+            var sources = sourcesBuilder.ToImmutableAndClear();
+            return AggregateSourcesIfNeeded(sources, isDocument);
+        }
+    }
+
+    public static ImmutableArray<IDiagnosticSource> AggregateSourcesIfNeeded(ImmutableArray<IDiagnosticSource> sources, bool isDocument)
+    {
+        if (sources.Length <= 1)
+        {
+            return sources;
+        }
+
+        if (isDocument)
+        {
+            // Group all document sources into a single source.
```

Comment:
i don't quite get the concept here.  why do we need to aggregate into a single source?

Reply:
We need to aggregate multiple sources when the client does not support multiple buckets - all the diagnostics (from all the sources) need to be reported in a single document report.

Reply:
VS code sends null `sourceName` for *.xaml files. It travels via PublicDocumentDiagnosticsHandler which expects [single IDiagnosticSource](https://github.com/dotnet/roslyn/blob/28b39facc518448c7af332286e847cefeb683d90/src/Features/LanguageServer/Protocol/Handler/Diagnostics/Public/PublicDocumentPullDiagnosticsHandler.cs#L84) which is in turn driven by https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#documentDiagnosticParams